### ISSUE TITLE
Adding plotting utilities

### ISSUE THREAD

**Issue by nhejazi** [benkeser/survtmle#5]
It's best to add plotting functions directly to `survtmle`, rather than having users produce their own plots. We'll use the standard plots for sieve analysis, for which code appears in the [`sieveml` repo](https://github.com/benkeser/sieveml).

It's best to port this code to `ggplot2`, both to have higher quality plots and to allow users to easily customize plots.

**Comment by nhejazi** [comment]
__code for plot layout setup:__ (can ignore when writing `ggplot` version)

```r
# figure dimensions -- used to compute appropriate margins
figure.width <- 8
figure.height <- 5.5
# computing appropriate margins
par(mar=c(4.1,0,0,0))
left.1 <- par()$mai[1]
par(mar=c(4.1,0,0,0))
right.1 <- par()$mai[1]
left.2 <- right.2 <- right.1
bottom.1 <- upper.2 <- right.1
bottom.2 <- upper.1 <- left.1
thisCex <- 1
x <- (figure.width - (left.1 + right.1 + left.2 + right.2))/2
wid <- rep(c(1,(x+left.2+right.2)/(x+left.1+right.1)),2)
x <- (figure.height - (upper.1 + upper.2 + bottom.1 + bottom.2))/2
hei <- c((x+upper.2+bottom.2)/(x+upper.1+bottom.1),1,1,(x+upper.2+bottom.2)/(x+upper.1+bottom.1))
```

**Comment by nhejazi** [comment]
__cumulative incidence plots:__

```r
# set up the layout
layout(matrix(c(1,3,2,4),2,2),width=wid)
# make the top row of plots
for(j in 1:2){
  if(j==2){
    par(mar=c(2.1,4.1,4.1,2.1))
    main.tit <- ""
    yl <- ""
  }else if(j==1){
    par(mar=c(2.1,4.1,4.1,2.1))
    main.tit <- ""
    yl <- "Cumulative incidence"
  }
  # get vector of cumulative incidence in vaccine arm
  f0.hat <- unlist(lapply(tp.fit, function(f){
      f$est[(2*j - 1)]
  }))
  # get vector of cumualtive incidence in vaccine arm
  f1.hat <- unlist(lapply(tp.fit, function(f){
      f$est[(2*j)]
  }))
  
  plot(c(0,f0.hat)~I(0:6),cex.lab=1.2,pch="",col=1,
       ylim=c(0,0.010),bty="n",cex=thisCex,
       #     ylim=c(0,0.004),
       ylab=yl,
       xaxt="n",xlab="Years since entry",
       main=main.tit)
  segments(x0=0,x1=7, y0=seq(0,0.010,0.002),col="gray90")
  points(c(0,f0.hat)~I(0:6),pch=ifelse(j==2,19,17),col=1,cex=thisCex)
  points(c(0,f1.hat)~I(0:6),pch=ifelse(j==2,1,2),col=1,cex=thisCex)
  axis(side=1, at=seq(0,6,2),label=seq(0,3,1))
  legend(x="topleft",pch=ifelse(rep(j,2)==c(2,2),c(19,1),c(17,2)),col=1,
         bty="n",
         legend=c("Placebo","Vaccine"))
  mtext(side=3,paste0("AA site 169",ifelse(j==1," matched"," mismatched")))
}
```

...I think we have these as of 84bae66

**Comment by nhejazi** [comment]
__vaccine efficacy plots:__

```r
#--------------------------
# vaccine efficacy plot
#--------------------------
# get matrix of vaccine efficacies, first column = matched
veMat <- Reduce("rbind", lapply(ve.fit,function(f){
    f$est[1:2]
}))

par(mar=c(4.1,4.1,2.1,2.1))
yl <- "Vaccine Efficacy"
y.lim <- c(-0.75,1.25)
plot(c(0,veMat[,1])~I(0:6),ylab=yl,col=1,
     xlab="Years since entry",cex.lab=1.2,xaxt="n",bty="n",cex=thisCex,
     ylim=y.lim,pch="")
segments(x0=0,x1=7, y0=seq(-1,1,0.5),col="gray90")
points(veMat[1:6,1]~I(1:6),ylab=yl,col=1,pch=17)
points(veMat[2:6,2]~I(2:6),ylab=yl,col=1,pch=19,lty=2)
axis(side=1, at=seq(0,6,2),label=seq(0,3,1))
legend(x=0,y=1.55,xpd=TRUE, pch=c(17,19),col=1,bty="n",
       legend=c("169 matched","169 mismatched"))
```

**Comment by nhejazi** [comment]
__sieve effect plot:__

```r
#-------------------------
# sieve effect plot
#-------------------------
# get vector of vse
vse <- Reduce("rbind", lapply(ve.fit,function(f){
    f["vse",1:3]
}))

par(mar=c(4.1,4.1,2.1,2.1))
yl <- "Sieve Effect"
y.lim <- c(0,12)

plot(c(0,0,vse[2:6,1])~I(0:6),ylab=yl,col=1,
     xlab="Years since entry",cex.lab=1.2,xaxt="n",bty="n",cex=thisCex,
     ylim=y.lim,pch="")
segments(x0=0,x1=7, y0=seq(0,10,2),col="gray90")
points(vse[2:6,1]~I(2:6),ylab=yl,col=1,pch=19)
axis(side=1, at=seq(0,6,2),label=seq(0,3,1))
abline(h=1,lty=3)
segments(x0=4:6,y0=vse[4:6,2],y1=vse[4:6,3])
arrows(x0=2:3,y0=vse[2:3,2],y1=12,length=0.075)
```

**Comment by benkeser** [comment]
A couple plot functions that would be good to have -- 

plot.timepoints
plotting method for timepoints object that would produce a plot of an incidence or a survival curve, i.e., 1 - incidence (as an option) for each trt level and each ftype (or a specified subset of them as an option) over time, with/without pointwise confidence intervals (as an option). An additional option to make a J panel plot by ftype would be nice (i.e., plot vaccine and placebo curves for each of J failure types). 

plot.survtmle
If survtmle is called at a fixed timepoint, then return a plot with the treatment effect (option for additive trt = 1 - trt = 0 OR multiplicative trt = 1 / trt = 0) and confidence interval (option, default = TRUE) for each ftype. This would allow you to visually assess sieve effects. Could also add options for displaying p-values for hypothesis tests (trt effect for each ftype and sieve effect test, i.e., are the trt effects the same across failure types -- this latter test would require some leg work to write for a general number of ftypes). I guess if this gets called and there's only one ftype, it's a rather boring plot.

**Comment by benkeser** [comment]
```
library(survtmle)
set.seed(341796)

n <- 100
t_0 <- 100
W <- data.frame(W1 = runif(n), W2 = rbinom(n, 1, 0.5))
A <- rbinom(n, 1, 0.5)
T <- rgeom(n,plogis(-4 + W$W1 * W$W2 - A)) + 1
C <- rgeom(n, plogis(-6 + W$W1)) + 1
ftime <- pmin(T, C)
ftype <- as.numeric(ftime == T)

fit1 <- survtmle(ftime = ftime, ftype = ftype,
                 adjustVars = W, glm.ftime = "I(W1*W2) + trt + t",
                 trt = A, glm.ctime = "W1 + t", method = "hazard",
                 verbose = TRUE,  t0 = t_0, maxIter = 2)

# fit at all timepoints
tpfit <- timepoints(fit1, times = seq_len(t_0))
```

Everything below this should be in `plot.tp.survtmle`...
```
grbg <- lapply(tpfit,function(x){x$est})
rslt <- Reduce(cbind,grbg)

plot(rslt[1,]~I(seq_len(t_0)), col = 1, type = "l")
points(rslt[2,]~I(seq_len(t_0)), col = 2, type = "l")

iso <- apply(rslt,1,function(y){
	tmp <- isoreg(y = y, x = seq_len(t_0))
	tmp$yf
})

lines(iso[,1]~I(seq_len(t_0)), col = 1, lty = 3)
lines(iso[,1]~I(seq_len(t_0)), col=1, lty = 3)
plot(iso[,1]~I(seq_len(t_0)), col = 1, lty = 3, type = "l")
lines(iso[,2]~I(seq_len(t_0)), col = 2, lty = 3)
```
n.b., use of "100" is from call to "t0" in `survtmle`, not "n" -- should be able to draw this from the `call` object that ought to be stored in the `survtmle` object

**Comment by nhejazi** [comment]
New plotting function for class `tp.survtmle` produces this for "raw" estimates (without isotonic regression to induce monotonicity):
<img width="741" alt="plot_timepoints-raw" src="https://user-images.githubusercontent.com/8527276/27892988-9a652ad8-61b7-11e7-89ec-abb751dde980.png">

...And here's the "smoothed" version, after the call to `stats::isoreg`:
<img width="743" alt="plot_timepoints-smooth-iso" src="https://user-images.githubusercontent.com/8527276/27893009-baf67608-61b7-11e7-9079-f5a2c08a93b9.png">

__n.b.,__ differences are very slight.

**Comment by benkeser** [comment]
Looks great! We don't really expect large differences -- asymptotically the estimates are the same. The isotonic regression just gives it a more familiar look to applied practitioners.

**Comment by nhejazi** [comment]
Thanks!

Got it -- that makes total sense. Though I definitely couldn't tell the difference between them for a while, had to run the two data frames through `all.equal` to make sure there actually was a difference...

**Comment by nhejazi** [comment]
__TODO:__

`plot.survtmle`:
If survtmle is called at a fixed timepoint, then return a plot with the treatment effect (option for additive trt = 1 - trt = 0 OR multiplicative trt = 1 / trt = 0) and confidence interval (option, default = TRUE) for each ftype. This would allow you to visually assess sieve effects. Could also add options for displaying p-values for hypothesis tests (trt effect for each ftype and sieve effect test, i.e., are the trt effects the same across failure types -- this latter test would require some leg work to write for a general number of ftypes). I guess if this gets called and there's only one ftype, it's a rather boring plot.

**Comment by benkeser** [comment]
Let's punt on this until we put out a separate sieve analysis package.

**Comment by benkeser** [comment]
It'd be nice if the plot function didn't interpolate between points, i.e., in the working example we're really only estimating cum inc at integer values, but the plot draws linearly interpolates between those points. Would maybe be better if it was drawn as a step function, i.e., constant on (0,1], (1,2], etc...

**Comment by nhejazi** [comment]
the simplest way to make the curves in the plot step functions is to use `geom_step` (instead of current use of `geom_line`)...here's the difference, using the example from `README.Rmd`:

__original plot (with interpolation):__
<img width="819" alt="screen shot 2017-07-07 at 9 00 09 am" src="https://user-images.githubusercontent.com/8527276/27966160-14d35682-62f3-11e7-9c51-e6ee45f9123c.png">

__new plot (as step functions):__
<img width="823" alt="screen shot 2017-07-07 at 9 01 50 am" src="https://user-images.githubusercontent.com/8527276/27966188-32056cf4-62f3-11e7-9661-f56de9a5a7bd.png">

Does the second one get closer to what you'd like to see? (we can also add options for both to the plot method)

**Comment by benkeser** [comment]
Yep, I think the second one is what we'd like

**Comment by nhejazi** [comment]
Ok, the second plot it is! Changes completed as of commit 91e4bc8.

I'm going to close this issue since the other plots we've discussed are specific to sieve analysis and will be implemented by the time we close #14. Feel free to re-open if this is incorrect.
